### DISCUSSION TITLE
How to get CurrentWorkingDirectory while inside a ICommandPredictor?

### DISCUSSION THREAD

**Discussion by AnderssonPeter** [PowerShell/PowerShell#16657] [Q&A] [answered]
Hi I'm writing a `ICommandPredictor` and I need to access the `CurrentWorkingDirectory` while inside the `ICommandPredictor.GetSuggestion`.

I have tried both `Directory.GetCurrentDirectory()` and `Environment.CurrentDirectory` and both of them reflect the Path of where the session was started not where it currently is.

So is there any way to get the correct one while inside a `ICommandPredictor`?

**Comment by vexx32** [comment]
It doesn't look like there's an easily accessible way to do that. @SeeminglyScience might have some ideas. @daxian-dbw would be the person to ask in the end though, as he wrote much of that code as far as I know.

**Reply by SeeminglyScience** [reply]
@daxian-dbw feel free to correct me here, but I believe this also falls under the category of needing current runspace access and therefore not supported.

**Reply by AnderssonPeter** [reply]
Ok, i guess i need to find a workaround then 😐, is there someway i can get a callback in the normal Runspace when the current working directory changes?

**Reply by AnderssonPeter** [reply]
After skimming the source code i found the following https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.commandinvocationintrinsics.locationchangedaction?view=powershellsdk-7.0.0#System_Management_Automation_CommandInvocationIntrinsics_LocationChangedAction

With some luck it might help me solve the problem!

**Reply by iSazonov** [reply]
> I believe this also falls under the category of needing current runspace access and therefore not supported.

Yes, CWD is a state per runspace. ICommandPredictor hasn't SessionState parameter but a class implemented the ICommandPredictor could have.

**Comment by AnderssonPeter** [comment, marked as answer]
There might be a better way but I solved it the following way
```C#
public class CurrentWorkingDirectoryProvider : ICurrentWorkingDirectoryProvider
{
    private volatile string currentWorkingDirectory;
    private readonly SessionState sessionState;

    public CurrentWorkingDirectoryProvider(SessionState sessionState)
    {
        this.sessionState = sessionState;
        currentWorkingDirectory = sessionState.Path.CurrentFileSystemLocation.ProviderPath;
        sessionState.InvokeCommand.LocationChangedAction += LocationChanged;
    }

    private void LocationChanged(object? sender, LocationChangedEventArgs e)
    {
        currentWorkingDirectory = sessionState.Path.CurrentFileSystemLocation.ProviderPath;
    }

    public string CurrentWorkingDirectory => currentWorkingDirectory;

    public void Dispose()
    {
        sessionState.InvokeCommand.LocationChangedAction -= LocationChanged;
    }
}
```
The `CurrentWorkingDirectoryProvider` is constructed in a normal cmdlet and injected into the constructor of my `ICommandPredictor`.

As the `currentWorkingDirectory` is volatile it should be threadsafe to read and write to it from multiple threads.

**Reply by daxian-dbw** [reply]
This looks great! Thanks for sharing it 😄
